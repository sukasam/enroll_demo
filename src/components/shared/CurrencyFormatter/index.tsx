import Spinner from "Components/shared/Spinner";
import T from "Components/shared/Translate";
import { useCountryConfig } from "Constants/countryConfig";
import { useTranslations } from "Contexts/translation";
import { ReactNode } from "react";

type CurrencyFormatterProps = {
    children: ReactNode;
    currency?: string;
    allowFree?: boolean;
};

export default function CurrencyFormatter(
    props: CurrencyFormatterProps
): JSX.Element {
    const { children, currency = null, allowFree = false } = props;
    const { country, language } = useTranslations();

    const countryConfig = useCountryConfig();
    const hiddenDecimal = !!countryConfig?.marketExceptions?.hiddenDecimal;
    const useGlobalFormatCurrency =
        !!countryConfig?.marketExceptions?.useGlobalFormatCurrency;
    const locale = ((): string => {
        if (!language || !country) return "en-US";
        if (useGlobalFormatCurrency) return "en-US";
        return `${language.toLowerCase()}-${country.toUpperCase()}`;
    })();

    if (children === undefined || children === null || children === "") {
        return <Spinner dark />;
    }

    const renderContent = (): JSX.Element => {
        const numericChildren =
            typeof children === "string"
                ? Number(children)
                : Number(String(children));
        if (Number.isNaN(numericChildren)) {
            return <Spinner dark />;
        }

        if (numericChildren === 0 && allowFree) {
            return <T>right_column_payment_shipping_price</T>;
        }

        const currentCurrency = currency || countryConfig?.currency || "USD";

        return (
            <>
                {Intl.NumberFormat(locale, {
                    style: "currency",
                    currency: currentCurrency,
                    minimumFractionDigits: hiddenDecimal ? 0 : 2,
                    maximumFractionDigits: hiddenDecimal ? 0 : 2
                }).format(numericChildren)}
            </>
        );
    };

    return renderContent();
}
