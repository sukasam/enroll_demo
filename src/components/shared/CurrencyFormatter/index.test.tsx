import { render } from "@testing-library/react";
import CurrencyFormatter from "./index";

// Mock the dependencies
jest.mock(
    "Components/shared/Spinner",
    () =>
        function MockSpinner(): JSX.Element {
            return <div data-testid="spinner">Spinner</div>;
        }
);

jest.mock(
    "Components/shared/Translate",
    () =>
        function MockTranslate({
            children
        }: {
            children: string;
        }): JSX.Element {
            return <span>{children}</span>;
        }
);

jest.mock("Constants/countryConfig", (): any => ({
    useCountryConfig: (): { currency: string } => ({ currency: "USD" })
}));

jest.mock("Contexts/translation", (): any => ({
    useTranslations: (): { country: string; language: string } => ({
        country: "US",
        language: "en"
    })
}));

describe("CurrencyFormatter", () => {
    it("should render spinner when children is undefined", () => {
        const { container } = render(
            <CurrencyFormatter>{undefined}</CurrencyFormatter>
        );
        expect(container.firstChild).toMatchSnapshot();
    });

    it("should render spinner when children is null", () => {
        const { container } = render(
            <CurrencyFormatter>{null}</CurrencyFormatter>
        );
        expect(container.firstChild).toMatchSnapshot();
    });

    it("should render spinner when children is NaN", () => {
        const { container } = render(
            <CurrencyFormatter>Not a number</CurrencyFormatter>
        );
        expect(container.firstChild).toMatchSnapshot();
    });

    it("should render 'Free' when children is 0 and allowFree is true", () => {
        const { container } = render(
            <CurrencyFormatter allowFree>{0}</CurrencyFormatter>
        );
        expect(container.firstChild).toMatchSnapshot();
    });

    it("should render formatted currency", () => {
        const { container } = render(
            <CurrencyFormatter>{1000}</CurrencyFormatter>
        );
        expect(container.firstChild).toMatchSnapshot();
    });

    it("should render formatted currency with custom currency", () => {
        const { container } = render(
            <CurrencyFormatter currency="EUR">{1000}</CurrencyFormatter>
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});
