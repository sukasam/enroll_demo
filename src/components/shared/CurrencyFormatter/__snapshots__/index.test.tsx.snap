// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyFormatter should render 'Free' when children is 0 and allowFree is true 1`] = `
<span>
  right_column_payment_shipping_price
</span>
`;

exports[`CurrencyFormatter should render formatted currency 1`] = `$1,000.00`;

exports[`CurrencyFormatter should render formatted currency with custom currency 1`] = `â‚¬1,000.00`;

exports[`CurrencyFormatter should render spinner when children is NaN 1`] = `
<div
  data-testid="spinner"
>
  Spinner
</div>
`;

exports[`CurrencyFormatter should render spinner when children is null 1`] = `
<div
  data-testid="spinner"
>
  Spinner
</div>
`;

exports[`CurrencyFormatter should render spinner when children is undefined 1`] = `
<div
  data-testid="spinner"
>
  Spinner
</div>
`;
